[#120] Hyperlink support

ghillairet opened the pull request:
This PR adds support for hyperlink detection inside a Swagger document.

Hyperlink supported:
- URL links, through default URLHyperlinkDetector
- JSON references inside current document
- Path parameters link to parameter definition
- Tags to schema definitions
- Required properties to property definition

tfesenko committed:
Merge pull request #2 from RepreZen/master

Update from the open-source repo

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Started to provide support for hyperlinks inside swagger documents. The source configuration has been updated to add 2 hyperlink detectors, URLHyperlinkDetector and SwaggerHyperlinkDetector. The first detector can detect URL like link and open them in the default browser. The second is the detector that will provide hyperlinks for other kinds of elements such as schema object, parameter, response or path.

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Add support for hyperlinks on path parameters.

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Add extension indicating that SwagEdit provides hyperlinks.

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Add hyperlink detection for json references.
Start better structuring HyperlinkDetector

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Rework implementation of HyperlinkDetector and update tests.

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Split the logic of previously SwaggerHyperlinkDetector into separate detectors dedicated to detect specific kinds of links, making it easier to reason about it and test.
We currently have 3 detectors
- DefinitionHyperlinkDetector handles detection of links to/inside schema definitions
- JsonReferenceHyperlinkDetector handles detection of all JSON references ($ref)
- PathParamHyperlinkDetector handles detection of links from path parameters

tfesenko commented on com.reprezen.swagedit/plugin.xml:
```diff
@@ -134,5 +134,22 @@
+         point="org.eclipse.ui.workbench.texteditor.hyperlinkDetectors">
+      <hyperlinkDetector
+            activate="true"
+            class="com.reprezen.swagedit.editor.SwaggerHyperlinkDetector"
```
`com.reprezen.swagedit.editor.SwaggerHyperlinkDetector` does not exist.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/SwaggerDocument.java:
```diff
@@ -148,8 +152,88 @@ public JsonNode getNodeForPath(String path) {
+	 * @return region under the path
+	 */
+	public IRegion getRegion(String path) {
+		if (Strings.emptyToNull(path) == null)
```
Inconsistent usage of the `{}` in the `if`s. You omit them here, but then use the 165-167. The code is good, but it makes it more difficult to read.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/SwaggerHyperlink.java:
```diff
@@ -0,0 +1,110 @@
+		builder.append(", target=");
+		builder.append(target);
+		builder.append("]");
+		return builder.toString();
```
Writing `toString()` is not the most exciting thing in life, you could have used `Objects.toStringHelper` from Guava, as in http://eclipsesource.com/blogs/2012/07/24/google-guava-quickie-clean-tostring-methods/ . And we already have a dependency to Guava, so it's OK.

ghillairet replied:
toString(), equals() and hashCode() are generated by the IDE, so not sure it's worth to use a library to handle that.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/SwaggerHyperlink.java:
```diff
@@ -0,0 +1,110 @@
+		} else if (!text.equals(other.text))
+			return false;
+
+		return true;
```
Guava `Objects` would be helpful here (and in hashCode()) as well, see https://gist.github.com/rocketraman/1399080

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/AbstractSwaggerHyperlinkDetector.java:
```diff
@@ -0,0 +1,104 @@
+
+		public final IRegion region;
+		public final String text;
+		public final int column;
```
I love final fields!

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/PathParamHyperlinkDetector.java:
```diff
@@ -0,0 +1,147 @@
+ * Contributors:
+ *    ModelSolv, Inc. - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+package com.reprezen.swagedit.editor;
```
We already have 5 classes related to hyperlinks, do you think it makes sense to move them into their own subpackage?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/PathParamHyperlinkDetector.java:
```diff
@@ -0,0 +1,147 @@
+	protected static final Pattern PARAMETER_PATTERN = Pattern.compile("\\{(\\w+)\\}");
+
+	@Override
+	public IHyperlink[] detectHyperlinks(ITextViewer textViewer, IRegion region, boolean canShowMultipleHyperlinks) {
```
All three existing subclasses of `AbstractSwaggerHyperlinkDetector` have the same structure which can be reused and, thus, make it DRYer:

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/PathParamHyperlinkDetector.java:
```diff
@@ -0,0 +1,147 @@
+			basePath = document.getPath(region);
+		} catch (BadLocationException e) {
+			basePath = null;
+		}
```
The all start with getting a basePath, it takes 7 lines. Can it be reused? Can the overall structure of `detectHyperlinks` be implemented in the super class which would provide hook methods for subclasses?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/PathParamHyperlinkDetector.java:
```diff
@@ -0,0 +1,147 @@
+		// not a path definition
+		if (emptyToNull(basePath) == null || !basePath.startsWith(":paths:")) {
+			return null;
+		}
```
The `basePath` is used to check if this particular HyperlinkDetector is applicable for the given `basePath`. This part has to be implemented in subclasses.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/PathParamHyperlinkDetector.java:
```diff
@@ -0,0 +1,147 @@
+
+		if (info == null) {
+			return null;
+		}
```
These 11 lines related retrieval of `HyperlinkInfo` info are the same in the subclasses

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/PathParamHyperlinkDetector.java:
```diff
@@ -0,0 +1,147 @@
+			}
+		}
+
+		return links.isEmpty() ? null : links.toArray(new IHyperlink[links.size()]);
```
The part where we actually calculate hyperlinks is also specific to a `HyperlinkDetector`

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/editor/DefinitionHyperlinkDetectorTest.java:
```diff
@@ -0,0 +1,88 @@
+				"  - foo\n" + 
+				"definitions:\n" + 
+				"  foo:\n" + 
+				"    type: object\n";
```
This would be a perfect use of the Xtend's multiline strings.

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/editor/PathParamHyperlinkDetectorTest.java:
```diff
@@ -0,0 +1,95 @@
+
+		// expected region
+		IRegion linkRegion = new Region(document.getLineOffset(1) + "  /".length(), "{id}".length());
+		IRegion targetRegion = new Region(document.getLineOffset(4), 86);
```
I like the approach with mocks you found to test `HyperlinkDetector`. Testing editors is hard, I am glad that you found a relatively easy way to test this.

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/editor/PathParamHyperlinkDetectorTest.java:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	@Test
+	public void testShouldCreateHyperLink_FromPathParameter_ToParameterDefinition() throws Exception {
```
I also like using the underscores (`_`) in test method names, it makes seeing the test results so much easier.

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Move classes related to hyperlink to their own package.
Refactor detectors to avoid duplication of code.

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Fix extension for hyperlinks

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Fix inconsistent formatting.

ghillairet committed:
[#120] Hyperlink support ("go to definition")

Add support for links to yaml files inside the workspace. Hyperlinks work for absolute and relative references.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/hyperlinks/PathParamHyperlinkDetector.java:
```diff
@@ -39,32 +39,12 @@
+	}
 
+	@Override
+	protected IHyperlink[] doDetect(SwaggerDocument doc, ITextViewer viewer, HyperlinkInfo info, String basePath) {
```
Thanks!

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/hyperlinks/JsonReferenceHyperlinkDetector.java:
```diff
@@ -37,42 +49,102 @@ protected boolean canDetect(String basePath) {
+		}
+	}
+
+	private IHyperlink[] doDetectExternalLink(HyperlinkInfo info) {
```
I modified the expanded petstore example with external references like this:
<img width="954" alt="screen shot 2016-05-18 at 4 33 11 pm" src="https://cloud.githubusercontent.com/assets/644582/15374215/23f68dc8-1d17-11e6-8b68-b6a37af07186.png">

and the navigation does not work. 

I understand that this is a corner case. What can I do to make it work?

ghillairet replied:
It will fail at this point https://github.com/RepreZen/SwagEdit/blob/task/120/com.reprezen.swagedit/src/com/reprezen/swagedit/editor/hyperlinks/JsonReferenceHyperlinkDetector.java#L120 because % is an illegal character for a URI.
